src/_pytest/python.py

```diff
@@ -911,6 +911,124 @@ def hasnew(obj: object) -> bool:
     return False
 
 
+@final
+@attr.s(frozen=True, auto_attribs=True)
+class IdMaker:
+    argnames: Iterable[str] = attr.Factory(list)
+    parametersets: Iterable[ParameterSet] = attr.Factory(list)
+    idfn: Optional[Callable[[Any], Optional[object]]] = None
+    ids: Optional[List[Union[None, str]]] = None
+    config: Optional[Config] = None
+    nodeid: Optional[str] = None
+
+    def make_unique_test_ids(self) -> List[str]:
+        resolved_ids = list(self._make_resolved_ids())
+
+        # All IDs must be unique!
+        unique_ids = set(resolved_ids)
+        if len(unique_ids) != len(resolved_ids):
+
+            # Record the number of occurrences of each test ID.
+            test_id_counts = Counter(resolved_ids)
+
+            # Map the test ID to its next suffix.
+            test_id_suffixes: Dict[str, int] = defaultdict(int)
+
+            # Suffix non-unique IDs to make them unique.
+            for index, test_id in enumerate(resolved_ids):
+                if test_id_counts[test_id] > 1:
+                    resolved_ids[index] = f"{test_id}{test_id_suffixes[test_id]}"
+                    test_id_suffixes[test_id] += 1
+
+        return resolved_ids
+
+    def _make_resolved_ids(self) -> Iterable[str]:
+        """Combine multiple parameter IDs to a single string for each parameter set."""
+        for valindex, parameterset in enumerate(self.parametersets):
+            if parameterset.id is not None:
+                yield parameterset.id
+            else:
+                id = self._id_at_index(valindex)
+                if id is None:
+                    yield self._test_id_from_parameterset(parameterset, valindex)
+                else:
+                    yield _ascii_escaped_by_config(id, self.config)
+
+    def _id_at_index(self, valindex: int) -> Optional[str]:
+        if self.ids and valindex < len(self.ids):
+            return self.ids[valindex]
+        return None
+
+    def _test_id_from_parameterset(self, parameterset, valindex) -> str:
+        this_id = [
+            self._idval(val, argname, valindex)
+            for val, argname in zip(parameterset.values, self.argnames)
+        ]
+        return "-".join(this_id)
+
+    def _evaluate_idval_function(
```

Comment:
Up to here the order of the functions in the class was `<function F>, <functions called by F>`. So I think it would be better to bring `def _idval` here before `_evaluate_idval_function` and the functions it calls.